front/lib/api/chat.ts

```diff
@@ -299,3 +327,335 @@ export async function updateChatMessageFeedback({
+  message: ChatMessageType;
+};
+
+// Event sent when receiving streamed response from the model.
+export type ChatMessageTokensEvent = {
+  messageId: string;
+  text: string;
+};
+
+// Event received when the session is updated (eg title is set).
+export type ChatSessionUpdateEvent = {
+  session: ChatSessionType;
+};
+
+/**
+ * Function that emulates starting a new chat session.
+ *
+ * @param auth Authenticator
+ * @param userMessage string
+ * @param dataSources list of data sources to use for retrieval
+ * @param filter filter to use for retrieval (timestamp)
+ * @param timeZone timezone to use for retrieval
+ */
+export async function* newChat(
+  auth: Authenticator,
+  userMessage: string,
+  dataSources:
+    | {
+        workspace_id: string;
+        data_source_id: string;
+      }[]
+    | null,
+  filter: {
+    timestamp: { gt: number };
+  } | null,
+  timeZone: string
+): AsyncGenerator<
+  | ChatSessionCreateEvent
+  | ChatMessageTriggerEvent
+  | ChatMessageCreateEvent
+  | ChatMessageTokensEvent
+  | ChatSessionUpdateEvent
+> {
+  const owner = auth.workspace();
+  if (!owner) {
+    throw new Error("Could not authenticate against the workspace.");
+  }
+
+  // create a Session
+  const sId = newChatSessionId();
+  const session = await upsertChatSession(auth, sId, null);
+
+  if (!session) {
+    // If we fail to create the session we simply yield error messages events and return.
+    yield { role: "error" } as ChatMessageTriggerEvent;
+    yield {
+      message: {
+        sId: new_id(),
+        role: "error",
+        message: "Could not create chat session.",
+      },
+    };
+    return;
+  }
+
+  yield { session } as ChatSessionCreateEvent;
+
+  const assistantConfig = cloneBaseConfig(
+    DustProdActionRegistry["chat-assistant-wfn"].config
+  );
+
+  const assistantContext = {
+    user: {
+      username: auth.user()?.username,
+      full_name: auth.user()?.name,
+    },
+    workspace: owner.name,
+    date_today: new Date().toISOString().split("T")[0],
+  };
+
+  const messages: ChatMessageType[] = [
+    {
+      sId: new_id(),
+      role: "user",
+      message: userMessage,
+    },
+  ];
+
+  // This is an helper function that adds an error message to the `messages` array and yield the
+  // proper events in case of error.
+  function* handleError(
+    errorMessage: string
+  ): Generator<ChatMessageTriggerEvent | ChatMessageCreateEvent> {
+    messages.push({
+      sId: new_id(),
+      role: "error",
+      message: errorMessage,
+    });
+
+    yield { role: "error" } as ChatMessageTriggerEvent;
+    yield {
+      message: messages[messages.length - 1],
+    } as ChatMessageCreateEvent;
+  }
+
+  // Master loop that will run until the last message is not a "retrieval" request. It will call the
+  // assistant and push to the `messages` array. The assistant either push an "assistant" message
+  // with a response or a "retrieval" message that will contain a query, if that's the case we run
```

Comment:
In front, at the time 2 retrievals are stored, one with the query & one with the docs
This was done with the idea that things happen in 2 steps, one with computing the params (query & later time, data sources, etc) and one with actually getting the docs , honestly more than default than conviction

We could switch to a one message logic, no strong convictions on this but would require a migration. We all love migrations :heart_decoration:

Reply:
I think it's a better model to have one message eventually :+1: 

I'll be happy to take that migration once we move the client to this logic :+1:

Reply:
Ah I get it this means we need a migration + change the client-side logic to map to this new logic because the conversations that will be created by this code will not be parseable by front

Reply:
Will the current client code be able to display this format? I guess yes no? In whic case we can do the migration later?